docs/logging.rst

```diff
@@ -65,3 +66,55 @@ The ``/login`` route adds the following attributes:
+(If not present, the connecting IP address will be used as the IP address of the remote client, but this almost certainly indicates Gafaelfawr has not been deployed correclty.)
+
+Making this work properly requires some additional configuration:
+
+#. Set the ``proxies`` configuration setting to the list of networks used for the NGINX ingress and any other proxies.
+   See :ref:`settings`.
+
+#. Configure the NGINX ingress to generate full, chained ``X-Forwarded-For`` headers.
+   Do this by adding:
+
+   .. code-block:: yaml
+
+      data:
+        compute-full-forwarded-for: "true"
+        use-forwarded-headers: "true"
+
+   to the ``ConfigMap`` for the NGINX ingress.
+   See the `NGINX Ingress Controller documentation <https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/configmap/>`__ for more details.
+   Be aware that this will affect every service behind the NGINX ingress, not just Gafaelfawr, so all other services must be prepared for receiving a full ``X-Forwaded-For`` header, possibly including (on the left-most end) addresses sent by a malicious client.
+   There is more information at `the Wikipedia article on X-Forwarded-For <https://en.wikipedia.org/wiki/X-Forwarded-For>`__.
+
+   This workaround would no longer be needed if `this feature request for the NGINX ingress were implemented <https://github.com/kubernetes/ingress-nginx/issues/5547>`__.
+
+#. Disable Kubernetes source IP NAT for the NGINX ingress.
+   This is required on GKE and may be required on other Kubernetes environments.
+   Do this by adding ``spec.externalTrafficPolicy`` to ``Local`` in the ``Service`` resource definition for the NGINX ingress controller.
+   This comes with some caveats and drawbacks.
+   See `this Medium post <https://medium.com/pablo-perez/k8s-externaltrafficpolicy-local-or-cluster-40b259a19404>`__ for more details.
+
+For the curious, here are the details of why these changes are required.
+
+Determining the client IP from ``X-Forwarded-For`` is complicated because Gafaelfawr's ``/auth`` route is called via an NGINX ``auth_request`` directive.
+In the Kubernetes NGINX ingress, this involves three layers of configuration.
+The protected service will have an ``auth_request`` directive that points to a generated internal location.
+That internal location will set ``X-Forwarded-For`` and then proxy to the ``/auth`` route.
+The ``/auth`` route configuration is itself a proxy that also sets ``X-Forwarded-For`` and then proxies the request to Gafaelfawr.
+Because of this three-layer configuration, if NGINX is configured to always replace the ``X-Forwarded-For`` header, Gafaelfawr will receive a header containing only the IP address of the NGINX ingress.
+
+The above configuration tells the NGINX ingress to instead retain the original ``X-Forwarded-For`` and append each subsequent client IP.
```

Comment:
I just noticed that I'm not sure this says what header the original host is emitted to?  Or are we not going to pass that back to people as well?

Reply:
Oh, that is added as `X-Original-Host`. I'll note that explicitly.

Reply:
`X-Forwarded-Host` as it turns out. But I think I misunderstood your note -- the hostname of the client isn't recorded or logged anywhere. `X-Forwarded-Host` is the `Host` header sent by the client and names the host of the server. That isn't logged by Gafaelfawr, although it does use it for the `/login` and `/logout` routes. I'll make a note of that.

Reply:
Sorry yeah this is the confusing part about which host - sorry I meant the client IP of the original machine making the request - which header is that emitted to?  Is that something a backend service will have access to to log itself as well? Or is it all just logged in gafaelfawr  and not passed on?

Not so worried about the original Virtual host field thing, which is the other Host field that I always get confused with :)

Reply:
Oh, the client IP of the original machine ends up in `X-Forwarded-For`. Gafaelfawr only parses it, and doesn't set the results of that parse into any supplemental header, but the protected service would receive the same `X-Forwarded-For` header and could in theory do the same thing.

Do you think we should set the results of Gafaelfawr's parsing in some other header that the backend could use more simply?

Reply:
Ah okay.  I think it's fine for now, I forgot that it would probably be there on the other side as well.  But sometimes frameworks also tend to mess this up and not deal with picking the right one, as you well know having fixed one.  It's probably fine for now, although I could maybe see just taking what we know to be the real IP and plopping that somewhere, but that's probably a future thing.